Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Plans.hs

```diff
@@ -0,0 +1,169 @@
+import qualified Storage.Queries.PlanDetails as QPD
+import Tools.Error
+import qualified Tools.Payment as Payment
+
+--- subscription ---
+
+--- To do clarify the amount calculactions and send appropriate amounts ---
+
+subcribe ::
+  (Id DP.Person, Id DM.Merchant) ->
+  Id PlanDetails ->
+  PaymentMode ->
+  Flow SubscriptionResponse
+subcribe (driverId, merchantId) planId paymentMode = do
+  plan <- runInReplica $ QPD.findByIdAndPaymentMode planId paymentMode >>= fromMaybeM (PlanNotFound planId.getId)
+  driverPlan <- runInReplica $ QDPlan.findByDriverIdAndPlanId driverId planId
+  runNoTransaction $ QDInfo.updatePlanPaymentMode paymentMode (cast driverId)
+  when (maybe False (\dp -> dp.planStatus == ACTIVE_PLAN) driverPlan) (throwError $ InvalidPlanStatus driverId.getId)
+  now <- getCurrentTime
+  driverRegisterationInvoice <- mkDriverFeeForMandate 1 driverId --- amount is rs. 1 needs to be verified ---
+  case plan.paymentMode of
+    AUTOPAY -> do
+      case driverPlan of
+        Just dPlan -> do
+          when (maybe True checkMandateStatus (dPlan.mandateStatus)) (throwError $ InvalidMandateStatus driverId.getId)
+          runNoTransaction $ QDF.create driverRegisterationInvoice
+          runNoTransaction $ QDPlan.updatePlanIdByDriverId driverId planId
+          order <- createMandateOrder (driverId, merchantId) driverRegisterationInvoice.id 1.0 --- amounts/fees need to be refactored these are just dummies for dev ---
+          return $ buildSubscriptionResponse (Just order)
+        Nothing -> do
+          runNoTransaction $ QDF.create driverRegisterationInvoice
+          newDriverPlan <- mkDriverPlanEntry planId driverId paymentMode INACTIVE_PLAN now 0 --- amounts/fees need to be refactored these are just dummies for dev ---
+          runNoTransaction $ QDPlan.create newDriverPlan
+          order <- createMandateOrder (driverId, merchantId) driverRegisterationInvoice.id 1.0 --- amounts/fees need to be refactored these are just dummies for dev ---
+          return $ buildSubscriptionResponse (Just order)
+    MANUAL -> do
+      case driverPlan of
+        Just dPlan -> do
+          when (maybe False checkMandateStatus (dPlan.mandateStatus)) (throwError $ InvalidMandateStatus driverId.getId)
+          runNoTransaction $ QDPlan.updatePlanIdByDriverId driverId planId
+          return $ buildSubscriptionResponse Nothing
+        Nothing -> do
+          newDriverPlan <- mkDriverPlanEntry planId driverId paymentMode INACTIVE_PLAN now 0
+          runNoTransaction $ QDPlan.create newDriverPlan
+          return $ buildSubscriptionResponse Nothing
+
+select ::
+  (Id DP.Person, Id DM.Merchant) ->
+  Id PlanDetails ->
+  PaymentMode ->
+  Flow AckResponse
+select (driverId, _) planId paymentMode = do
+  driverInfo <- runInReplica $ QDInfo.findById (cast driverId) >>= fromMaybeM DriverInfoNotFound
+  currentDriverPlan <- runInReplica $ QDPlan.findByDriverId driverId >>= fromMaybeM (NoCurrentPlanForDriver driverId.getId)
```

Comment:
Why are we throwing error here in these cases? Why are we not allowed to select a plan without having an existing/Inactive plan?